# Stepwise conditional analysis. Configuration: Deletion-only.

#################################################
### Libraries & External arguments ##############
library(data.table)
library(dplyr)
library(tidyr)
library(tibble)
library(stringr)
library(gtools)

# Arguments
args <- commandArgs(trailingOnly = T)
sex <- args[1]
print(paste0("Analyzing: ", sex))



#################################################
### SET UP WHILE LOOP ###########################

round <- 1

# Number of significant associations; File generated by "05_gwas/05_pruning/deletion_only/01_pruning_del.R" 
n <- nrow(as.data.frame(fread(paste0("/cauwerx/gwas/05_gwas/05_pruning/deletion_only/data/", sex, "/00_qc_probes_gwas_del_", sex, ".txt"), header = T)))


while (n > 0) {
print(paste0("STARTING ROUND ", round))



#################################################
### STEP 1: Select top probes ###################

# Access the new significant data
if (round == 1) {
		# At first round, file generated by "05_gwas/05_pruning/deletion_only/01_pruning_del.R"
		new_sig <- as.data.frame(fread(paste0("/cauwerx/gwas/05_gwas/05_pruning/deletion_only/data/", sex, "/00_qc_probes_gwas_del_", sex, ".txt"), header = T))
	} else {
		# At next rounds, file generated by this script	
		new_sig <- as.data.frame(fread(paste0("/cauwerx/gwas/06_SCA/deletion_only/data/temp/significant_associations/", sex, "/sig_assoc_binary_del_SCA_", sex, "_round", round-1, ".txt.gz"), header = T))}

# Extract top variant/disease
new_sig <- new_sig[order(new_sig$P), ]
new_sig <- new_sig[!duplicated(new_sig[, "PHENO"]), ]
new_sig$ROUND <- round
print(paste0("Number of phenotypes with at least one significant association added at round ", round,": ", nrow(new_sig)))



#################################################
### STEP 2: Fill summary file ###################

# Create (round 1) or load (from previous round, file generated by this script) the summary file
if (round == 1) {
		sum_file <- new_sig
	} else {
		sum_file <- as.data.frame(fread(paste0("/scratch/gwas/06_SCA/deletion_only/data/temp/summary/", sex, "/summary_del_SCA_", sex, "_round", round-1, ".txt"), header = T))
		sum_file <- rbind(sum_file, new_sig)}

print(paste0("Total number of independent signals after round ",round,": ", nrow(sum_file)))
fwrite(sum_file, paste0("/cauwerx/gwas/06_SCA/deletion_only/data/temp/summary/", sex, "/summary_del_SCA_", sex, "_round", round, ".txt"), col.names = T, row.names = F, quote = F, sep = "\t", na = NA)



#################################################
### STEP 3: Generate probe CNV profile ##########

# Detect the probes whose CNV profile is to be extracted
probes <- sum_file[order(sum_file$ROUND), c(1:3, 17:18)]
probes <- probes[!duplicated(probes[, "ID"]), ]
probes <- probes[which(probes$ROUND == round), ]
print(paste0("Number of probes to prepare: ", nrow(probes)))

# Loop over probes to generate their CNV profile
if(nrow(probes) > 0) {
for(i in 1:nrow(probes)) {

	# Define probe
	rs <- as.character(probes[i, "ID"])
	chr <- as.character(probes[i, "#CHROM"])
	print(paste0("Starting probe: ", rs, " (chr", chr, ")"))

	# Write file
	fwrite(data.frame(rs), paste0("/cauwerx/gwas/06_SCA/deletion_only/data/temp/cnv_profiles/", sex, "/", rs), col.names = F, row.names = F, quote = F, sep = "\t")

	# Extract profile with PLINK
	input <- paste0("/data/UKBB/cnv_calls/PLINK/deletion_only/ukb_cnv_bed/ukb_cnv_chr", chr) # PLINK file set with deletion encoding; Generation of this file is described in Auwerx et al., 2022
	output <- paste0("/cauwerx/gwas/06_SCA/deletion_only/data/temp/cnv_profiles/", sex, "/", rs)
	samples <- paste0("/cauwerx/gwas/01_samples/data/samples_white_british_plink1.9_", sex, ".txt") # File generated by "01_samples/sample_filtering.R"
	system(paste("/cauwerx/softwares/plink/plink",
				 "--fam /data/UKBB/cnv_calls/PLINK/mirror/ukb_cnv_bed/ukb_cnv_chrX_onlyF.fam", # Fake .fam file (all individuals = females) to deal with PLINK's male hemizygozity assumption (see Supplemental Note 1) 
				 "--threads 20",
				 "--bfile", input, 
				 "--keep", samples,
				 "--extract", output,
				 "--recode tab --out", output))

	# Correct the .ped in R
 	ped <- as.data.frame(fread(paste0(output, ".ped"), header = F, select = c(2,7), col.names = c("IID", "CNV")))
	ped[which(ped$CNV == "0 0"), "CNV"] <- NA
	ped[which(ped$CNV == "A T" | ped$CNV == "T A"), "CNV"] <- 0
	ped[which(ped$CNV == "T T"), "CNV"] <- 1
	colnames(ped)[2] <- rs
	fwrite(ped, paste0(output, "_profile"), col.names = T, row.names = F, quote = F, sep = "\t")

	# Delete the rs file and the .ped/.map
	unlink(output); unlink(paste0(output, ".log")); unlink(paste0(output, ".map")); unlink(paste0(output, ".ped"))

}
}
rm(i, rs, chr, input, output, samples, ped)



#################################################
### STEP 4: Prepare covariates ##################
print("Starting to update the covariates")

# Loop over diseases
for (p in new_sig$PHENO) {

	# Define the variants to correct for
	new_cov <- new_sig[new_sig$PHENO == p, "ID"]
	print(paste0("Adding covariate to ", p,": ", new_cov, " (chr", new_sig[new_sig$PHENO == p, "#CHROM"], ":", new_sig[new_sig$PHENO == p, "POS"], ")"))

	# Load previous covariate file
	if (round == 1) {

			# Load full original covariate file; Generated by "03_covariates/01_covariate_extraction.R" 
			cov <- as.data.frame(fread("/cauwerx/gwas/03_covariates/data/covariates.txt", header = T))

			# Load disease-specific selected covariates (generated by "05_gwas/02_covariate_selection/01_covariate_selection.R"), if there is at least one, otherwise retain IID only
			if (file.size(paste0("/cauwerx/gwas/05_gwas/02_covariate_selection/data/", sex, "/covariates_", sex, ".", p,".txt")) > 1) {
				selected_cov <- as.data.frame(fread(paste0("/cauwerx/gwas/05_gwas/02_covariate_selection/data/", sex, "/covariates_", sex, ".", p,".txt"), header = F, sep = "\t"))
				selected_cov <- unlist(strsplit(selected_cov[1,1],","))
				# Retain selected covariates
				cov <- cov[, names(cov) %in% c("IID", selected_cov)]
			} else {
				# Retain the ID
				cov <- cov[, names(cov) %in% c("IID"), drop = F]}

	} else {
			# Load covariate file from previous round (generated by this script)
			cov <- as.data.frame(fread(paste0("/cauwerx/gwas/06_SCA/deletion_only/data/temp/covariates/", sex, "/covariates_", sex, "_round", round-1, ".", p,".txt"), header = T))}

	# Add the profile of the top probe to the covariate file
	new_cov_profile <- as.data.frame(fread(paste0("/cauwerx/gwas/06_SCA/deletion_only/data/temp/cnv_profiles/", sex, "/", new_cov,"_profile"), header = T))
	cov <- right_join(cov, new_cov_profile, by = "IID")

	# Save the updated covariate file & delete the old one
	fwrite(cov, paste0("/cauwerx/gwas/06_SCA/deletion_only/data/temp/covariates/", sex, "/covariates_", sex, "_round", round, ".", p,".txt"), col.names = T, row.names = F, quote = F, sep = "\t", na = NA)
	if(round != 1){ 
	unlink(paste0("/cauwerx/gwas/06_SCA/deletion_only/data/temp/covariates/", sex, "/covariates_", sex, "_round", round-1, ".", p,".txt"))}

}
rm(p, new_cov, new_cov_profile, cov)



#################################################
### STEP 5: GWAS ################################

# Note: GWAS will be performed by chromosome, recursively
print("Strating GWAS")

# Create a new directory to store results of this round
system(paste0("mkdir /cauwerx/gwas/06_SCA/deletion_only/data/temp/associations/", sex, "/round", round))

# List chromosomes
chromosomes <- c(1:22, "X", "XY")

# Loop over diseases
for (p in new_sig$PHENO) {

	for(chr in chromosomes) {

		# Define variables
		input <- paste0("/data/UKBB/cnv_calls/PLINK/deletion_only/ukb_cnv_bed/ukb_cnv_chr", chr) # PLINK file set with deletion encoding; Generation of this file is described in Auwerx et al., 2022
		pheno_file <- paste0("/cauwerx/gwas/04_phenotypes/data/pheno_ICD10_", sex, ".txt") # File generated by "04_phenotypes/01_ICD10_extraction.R"
		covar_file <- paste0("/cauwerx/gwas/06_SCA/deletion_only/data/temp/covariates/", sex, "/covariates_", sex, "_round", round, ".", p,".txt")	
		sample_file <- paste0("/cauwerx/gwas/01_samples/data/samples_white_british_", sex, ".txt") # File generated by "01_samples/sample_filtering.R"
		probe_file <- paste0("/cauwerx/gwas/05_gwas/05_pruning/deletion_only/data/", sex, "/SCA_probes/qc_probes_del_", sex, ".", p,".txt") # File generated by "05_gwas/05_pruing/deletion_only/01_pruning_del.R"
		output <- paste0("/cauwerx/gwas/06_SCA/deletion_only/data/temp/associations/", sex, "/round", round,"/gwas_binary_del_SCA_", sex, "_round", round,"_chr", chr)

		# GWAS in PLINK v2
		system(paste("/cauwerx/softwares/plink2/plink2",
					  "--threads 20",  	
					  "--bfile", input, 
					  "--fam /data/UKBB/cnv_calls/PLINK/mirror/ukb_cnv_bed/ukb_cnv_chrX_onlyF.fam", # Fake .fam file (all individuals = females) to deal with PLINK's male hemizygozity assumption (see Supplemental Note 1)
					  "--pheno iid-only",  pheno_file, "--no-psam-pheno", "--pheno-name", p,
					  "--covar iid-only", covar_file, "--covar-variance-standardize",
					  "--glm firth-fallback omit-ref no-x-sex hide-covar --ci 0.95",
					  "--keep", sample_file, 
					  "--extract", probe_file,
					  "--out", output))
	}
}
rm(chromosomes, chr, p, input, pheno_file, covar_file, sample_file, probe_file, output)



#################################################
### STEP 6: Correct GWAS & ID significants ######

# Threshold for significance
thr <- 0.05/6633

# List files to corrrect; Generated at previous step 5
list_gwas <- mixedsort(list.files(paste0("/cauwerx/gwas/06_SCA/deletion_only/data/temp/associations/", sex, "/round", round), pattern = ".glm", full.names = T, recursive = F))
print(paste0("Number of files to analyze: ", length(list_gwas)))

# Correct & extract significant signals
sig_gwas <- data.frame()

# Loop over diseases
for (f in list_gwas) {

	# Define the disease
	p <- sub(".*\\.", "",sub(".glm.*", "", f)) 

	# Load file
	df <- as.data.frame(fread(f, header = T))

	#### Continuous trait (DiseaseBurden) ####
	if (grepl("DiseaseBurden", f))  {

		# Make relevant columns numeric
		df[9:14] <- lapply(df[9:14], as.numeric)

		# Correct associations where A1 = "A"; Correct Beta, CI, Z, alleles 
		df[which(df$A1 == "A"), c(9, 11, 12, 13)] <- -1 * df[which(df$A1 == "A"), c(9, 11, 12, 13)]
		df[which(df$A1 == "A"), "REF"] <- "A"
		df[which(df$A1 == "A"), "ALT"] <- "T"
		df[which(df$A1 == "A"), "A1"] <- "T"
		colnames(df)[c(11,12)] <- c("U95", "L95")
		df <- df[, c(1:10, 12, 11, 13:15)]

		# Add a "FIRTH?" column forr compatibility
		df <- add_column(df, `FIRTH?` = NA, .before = "TEST")
		colnames(df)[c(10,11,14)] <- c("OR", "LOG(OR)_SE", "Z_STAT") 
                                                                                                              
	#### Binary traits (others) ##############
	} else {
	df[10:15] <- lapply(df[10:15], as.numeric)

	# Correct associations where A1 = "A"; Correct: OR, CI, Z, alleles; Not correct: SE(ln(OR)), P
	df[df$A1 == "A", "OR"] <- exp(-log(df[df$A1 == "A", "OR"]))
	df[df$A1 == "A", "L95"] <- exp(log(df[df$A1 == "A", "OR"]) - 1.96*df[df$A1 == "A", "LOG(OR)_SE"]) 
	df[df$A1 == "A", "U95"] <- exp(log(df[df$A1 == "A", "OR"]) + 1.96*df[df$A1 == "A", "LOG(OR)_SE"]) 
	df[df$A1 == "A", "Z_STAT"] <- log(df[df$A1 == "A", "OR"])/df[df$A1 == "A", "LOG(OR)_SE"]
	df[df$A1 == "A", "REF"] <- "A"
	df[df$A1 == "A", "ALT"] <- "T"
	df[df$A1 == "A", "A1"] <- "T"

	}
	
	# Save
	fwrite(df, f, col.names = T, row.names = F, quote = F, sep = "\t", na = NA)

	# Extract significant signals
	df$PHENO <- p
	temp_sig <- df[df$P <= thr & !is.na(df$P), ]
	sig_gwas <- rbind(sig_gwas, temp_sig)
}

print(paste0("Number of significant associations after round ", round, ": ", nrow(sig_gwas)))
fwrite(sig_gwas, paste0("/cauwerx/gwas/06_SCA/deletion_only/data/temp/significant_associations/", sex, "/sig_assoc_binary_del_SCA_", sex, "_round", round, ".txt.gz"), col.names = T, row.names = F, quote = F, sep = "\t", na = NA)
rm(f, df, temp_sig)



#################################################
### STEP 7: Update loop variables ###############

round <- round + 1
n <- nrow(sig_gwas)
rm(sig_gwas)
}
