# Locus zoom plot for the association between the 15q13 locus and red blood cell count according to the mirror association model.

#################################################
### Libraries & External arguments ##############
library(data.table)
library(dplyr)
library(tidyr)
library(stringr)

# Arguments
args <- commandArgs(trailingOnly = TRUE)
chr <- args[1]
start <- args[2]
end <- args[3]
pheno <- gsub("-", " ", args[4])
print(paste0("Locus zoom plot for: chr", chr, ":", start, "-", end, " - ", pheno))



#################################################
### STEP 1: High missingness file ###############

# Probes with > 95% missingness; File generated as described in Auwerx et al., 2022 ("GWAS/02_probes/frequency/02_high_missingness_0.95.R")
hm <- as.data.frame(fread("/data/UKBB/general_data/high_missingness/genotype_missingness_0.95.txt"))
fwrite(data.frame(hm$SNP), "/cauwerx/gwas/10_candidates/15q13/data/temp/locus_zoom/high_missingness.txt", col.names = F, row.names = F, quote = F, sep = "\t")



#################################################
### STEP 2: PLINK input #########################

# PLINK file set with mirror encoding; Generation of this file is described in Auwerx et al., 2022
plink_file <- paste0("/data/UKBB/cnv_calls/PLINK/mirror/ukb_cnv_bed/ukb_cnv_chr", chr)

# Phenotypes from Auwerx et al., 2022 (see "GWAS/04_phenotypes/continuous/phenotype_extraction.R")
pheno_file <- "/data/continuous_CNV_GWAS/pheno_continuous_WB_INT_age_age2_sex_batch_PCs_All.txt"

# Samples: white-british individuals; Generated by "01_samples/sample_filtering.R"
samples_file <- "/cauwerx/gwas/01_samples/data/samples_white_british_All.txt"

# Probes to exclude due to high missingness (step 1)
HM_file <- "/cauwerx/gwas/10_candidates/15q13/data/temp/locus_zoom/high_missingness.txt"

# Output file
output_file <- paste0("/cauwerx/gwas/10_candidates/15q13/data/temp/locus_zoom/mirror_All_chr", chr, ":", start, ":", end)



#################################################
### STEP 3:  Run PLINK ##########################

system(paste("/cauwerx/softwares/plink2/plink2",
 			 "--bfile", plink_file,
			 "--pheno",  pheno_file, "--no-psam-pheno", "--pheno-name", pheno,
			 "--glm omit-ref no-x-sex hide-covar allow-no-covars --ci 0.95", 
			 "--keep", samples_file,
			 "--exclude", HM_file, "--chr", chr, "--from-bp", start, "--to-bp", end,
			 "--out", output_file))
unlink(HM_file)
unlink(paste0(output_file, ".log"))



#################################################
### STEP 4: Correct associations ################

# Loop over files
for(p in unlist(strsplit(pheno, " "))) {

	# Load file
	df <- as.data.frame(fread(paste0(output_file, ".", p, ".glm.linear"), header = T))

	# Correct
	df[which(df$A1 == "A"), c(9, 11, 12, 13)] <- -1 * df[which(df$A1 == "A"), c(9, 11, 12, 13)]
	df[which(df$A1 == "A"), "REF"] <- "A"
	df[which(df$A1 == "A"), "ALT"] <- "T"
	df[which(df$A1 == "A"), "A1"] <- "T"
	colnames(df)[c(11,12)] <- c("U95", "L95")
	df <- df[, c(1:10, 12, 11, 13:15)]
	
	# Save
	fwrite(df, paste0("/cauwerx/gwas/10_candidates/15q13/data/locus_zoom/mirror_All_chr", chr, ":", start, "-", end, ".", p, ".glm.linear"), col.names = T, row.names = F, quote = F, sep = "\t", na = NA)

	# Delete old file
	unlink(paste0(output_file, ".", p, ".glm.linear"))
}
