# Stepwise conditional analysis. Configuration: Mirror. Note: GWAS in R to force usage of logistic (vs. Firth) regression. 

#################################################
### Libraries & External arguments ##############
library(data.table)
library(dplyr)
library(tidyr)
library(stringr)
library(gtools)

# Arrguments
args <- commandArgs(trailingOnly = T)
sex <- args[1]
print(paste0("Analyzing: ", sex))



#################################################
### SET UP WHILE LOOP ###########################

round <- 1

# Number of significant associations; File generated by "05_gwas/05_pruning/mirror/01_pruning_mirror.R" 
n <- nrow(as.data.frame(fread(paste0("/cauwerx/gwas/05_gwas/05_pruning/mirror/data/", sex, "/00_qc_probes_gwas_mirror_", sex, ".txt"), header = T)))


while (n > 0) {
print(paste0("STARTING ROUND ", round))



#################################################
### STEP 1: Select top probes ###################

# Access the new significant data
if (round == 1) {
		# At first round, file generated by "05_gwas/05_pruning/mirror/01_pruning_mirror.R"
		new_sig <- as.data.frame(fread(paste0("/cauwerx/gwas/05_gwas/05_pruning/mirror/data/", sex, "/00_qc_probes_gwas_mirror_", sex, ".txt"), header = T))
	} else {
		# At next rounds, file generated by this script	
		new_sig <- as.data.frame(fread(paste0("/cauwerx/gwas/06_SCA/mirror/data/temp/significant_associations/", sex, "/sig_assoc_binary_mirror_SCA_", sex, "_round", round-1, ".txt.gz"), header = T))}

# Extract top variant/disease
new_sig <- new_sig[order(new_sig$P), ]
new_sig <- new_sig[!duplicated(new_sig[, "PHENO"]), ]
new_sig$ROUND <- round
print(paste0("Number of phenotypes with at least one significant association added at round ", round,": ", nrow(new_sig)))



#################################################
### STEP 2: Fill summary file ###################

# Create (round 1) or load (from previous round, file generated by this script) the summary file
if (round == 1) {
		sum_file <- new_sig
	} else {
		sum_file <- as.data.frame(fread(paste0("/cauwerx/gwas/06_SCA/mirror/data/temp/summary/", sex, "/summary_mirror_SCA_", sex, "_round", round-1, ".txt"), header = T))
		sum_file <- rbind(sum_file, new_sig)}

print(paste0("Total number of independent signals after round ",round,": ", nrow(sum_file)))
fwrite(sum_file, paste0("/cauwerx/gwas/06_SCA/mirror/data/temp/summary/", sex, "/summary_mirror_SCA_", sex, "_round", round, ".txt"), col.names = T, row.names = F, quote = F, sep = "\t", na = NA)



#################################################
### STEP 3: Probe CNV profile ###################

# Detect the probes whose CNV profile is to be extracted
probes <- sum_file[order(sum_file$ROUND), c(1:3, 17:18)]
probes <- probes[!duplicated(probes[, "ID"]), ]
probes <- probes[which(probes$ROUND == round), ]
print(paste0("Number of probes to prepare: ", nrow(probes)))

# Loop over probes to generate their CNV profile
if(nrow(probes) > 0) {
for(i in 1:nrow(probes)) {

	# Define probe
	rs <- as.character(probes[i, "ID"])
	chr <- as.character(probes[i, "#CHROM"])
	print(paste0("Starting probe: ", rs, " (chr", chr, ")"))

	# Write file
	fwrite(data.frame(rs), paste0("/cauwerx/gwas/06_SCA/mirror/data/temp/cnv_profiles/", sex, "/", rs), col.names = F, row.names = F, quote = F, sep = "\t")

	# Extract profile with PLINK
	input <- paste0("/data/UKBB/cnv_calls/PLINK/mirror/ukb_cnv_bed/ukb_cnv_chr", chr) # PLINK file set with mirror encoding; Generation of this file is described in Auwerx et al., 2022
	output <- paste0("/cauwerx/gwas/06_SCA/mirror/data/temp/cnv_profiles/", sex, "/", rs)
	samples <- paste0("/cauwerx/gwas/01_samples/data/samples_white_british_plink1.9_", sex, ".txt") # File generated by "01_samples/sample_filtering.R"
	system(paste("/cauwerx/softwares/plink/plink",
				 "--fam /data/UKBB/cnv_calls/PLINK/mirror/ukb_cnv_bed/ukb_cnv_chrX_onlyF.fam", # Fake .fam file (all individuals = females) to deal with PLINK's male hemizygozity assumption (see Supplemental Note 1) 
				 "--threads 20",
				 "--bfile", input, 
				 "--keep", samples,
				 "--extract", output,
				 "--recode tab --out", output))

	# Correct the .ped in R
 	ped <- as.data.frame(fread(paste0(output, ".ped"), header = F, select = c(2,7), col.names = c("IID", "CNV")))
	ped[which(ped$CNV == "A A"), "CNV"] <- -1
	ped[which(ped$CNV == "A T" | ped$CNV == "T A"), "CNV"] <- 0
	ped[which(ped$CNV == "T T"), "CNV"] <- 1
	colnames(ped)[2] <- rs
	fwrite(ped, paste0(output, "_profile"), col.names = T, row.names = F, quote = F, sep = "\t")

	# Delete the rs file and the .ped/.map
	unlink(output); unlink(paste0(output, ".log")); unlink(paste0(output, ".map")); unlink(paste0(output, ".ped"))

}
}
rm(i, rs, chr, input, output, samples, ped, probes)



#################################################
### STEP 4: Prepare covariates ##################
print("Starting to update the covariates")

# Loop over diseases
for (p in new_sig$PHENO) {

	# Define the variants to correct for
	new_cov <- new_sig[new_sig$PHENO == p, "ID"]
	print(paste0("Adding covariate to ", p,": ", new_cov, " (chr", new_sig[new_sig$PHENO == p, "#CHROM"], ":", new_sig[new_sig$PHENO == p, "POS"], ")"))

	# Load previous covariate file
	if (round == 1) {

			# Load full original covariate file; Generated by "03_covariates/01_covariate_extraction.R" 
			cov <- as.data.frame(fread("/cauwerx/gwas/03_covariates/data/covariates.txt", header = T))

			# Load disease-specific selected covariates (generated by "05_gwas/02_covariate_selection/01_covariate_selection.R"), if there is at least one, otherwise retain IID only
			if (file.size(paste0("/cauwerx/gwas/05_gwas/02_covariate_selection/data/", sex, "/covariates_", sex, ".", p,".txt")) > 1) {
				selected_cov <- as.data.frame(fread(paste0("/cauwerx/gwas/05_gwas/02_covariate_selection/data/", sex, "/covariates_", sex, ".", p,".txt"), header = F, sep = "\t"))
				selected_cov <- unlist(strsplit(selected_cov[1,1],","))
				# Retain selected covariates
				cov <- cov[, names(cov) %in% c("IID", selected_cov)]
			} else {
				# Retain the ID
				cov <- cov[, names(cov) %in% c("IID"), drop = F]}

	} else {
			# Load covariate file from previous round (generated by this script)
			cov <- as.data.frame(fread(paste0("/cauwerx/gwas/06_SCA/mirror/data/temp/covariates/", sex, "/covariates_", sex, "_round", round-1, ".", p,".txt"), header = T))}

	# Add the profile of the top probe to the covariate file
	new_cov_profile <- as.data.frame(fread(paste0("/cauwerx/gwas/06_SCA/mirror/data/temp/cnv_profiles/", sex, "/", new_cov,"_profile"), header = T))
	cov <- right_join(cov, new_cov_profile, by = "IID")

	# Save the updated covariate file & delete the old one
	fwrite(cov, paste0("/cauwerx/gwas/06_SCA/mirror/data/temp/covariates/", sex, "/covariates_", sex, "_round", round, ".", p,".txt"), col.names = T, row.names = F, quote = F, sep = "\t", na = NA)
	if(round != 1){ 
	unlink(paste0("/cauwerx/gwas/06_SCA/mirror/data/temp/covariates/", sex, "/covariates_", sex, "_round", round-1, ".", p,".txt"))}

}
rm(p, new_cov, new_cov_profile, cov)



#################################################
### STEP 5: GWAS ################################

# Targeted association studies to identify independently associated loci 
print("Strating GWAS")

# Load probes (to match rs ID and genomic location); This file contains the rs ID and position of all probes on the genotype array, available from the UKBB portal
probe_index <- as.data.frame(fread("/data/general_data/UKBB/probes.txt.gz"))
probe_index$Chr <- gsub("23", "XY", probe_index$Chr)

# Create an empty dataframe to store association signals
gwas_results <- data.frame()

# Loop over diseases
count <- 1
for (p in new_sig$PHENO) {

	# Load diseases data
	df_pheno <- as.data.frame(fread(paste0("/cauwerx/gwas/04_phenotypes/data/pheno_ICD10_", sex, ".txt"))) # File generated by "04_phenotypes/01_ICD10_extraction.R"
	df_pheno <- na.omit(df_pheno[, names(df_pheno) %in% c("IID", p)])
	colnames(df_pheno)[2] <- "PHENO"
	if(p == "DiseaseBurden") {
		df_pheno$PHENO <- as.numeric(df_pheno$PHENO)
	} else {
		df_pheno$PHENO <- factor(df_pheno$PHENO, levels = c(1,2), labels = c(0, 1))
	}

	# Load covariate data; File generated at previous step (4)
	df_cov <- as.data.frame(fread(paste0("/cauwerx/gwas/06_SCA/mirror/data/temp/covariates/", sex, "/covariates_", sex, "_round", round, ".", p,".txt")))
	df_pheno_cov <- left_join(df_pheno, df_cov, by = "IID"); rm(df_pheno, df_cov)

	# Load probes to analyze, remove those included as covariates; File generated by "05_gwas/05_pruning/mirror/01_pruning_mirror.R"
	df_probes <- as.data.frame(fread(paste0("/cauwerx/gwas/05_gwas/05_pruning/mirror/data/", sex, "/SCA_probes/qc_probes_mirror_", sex, ".", p,".txt"), header = F, col.names = c("ID"))) 
	df_probes <- df_probes[!df_probes$ID %in% names(df_pheno_cov), , drop = F]

	# Loop over probes
	for (rs in df_probes$ID) {

		# Load probe; Files generate by "05_gwas/05_pruning/mirror/02_probe_profile_extraction.R"
		df_rs <- as.data.frame(fread(paste0("/cauwerx/gwas/05_gwas/05_pruning/mirror/data/probe_profiles/", rs,"_profile"), col.names = c("IID", "CNV")))

		# Merge with the phenotype and covariate dataframe and re-order (PHENO, CNV, COV(s))
		df <- left_join(df_pheno_cov, df_rs, by = "IID"); rm(df_rs)
		df <- df[, c(2, ncol(df), 3:(ncol(df)-1))]

		# If analyzing "DiseaseBurden", perform a linear regression		
		if (p == "DiseaseBurden") {
			fit <- lm(PHENO ~ . , data = df)
			CI <- confint(fit)

			# Fill in the result table following the PLINK scheme
			gwas_results[count, "#CHROM"] <- probe_index[probe_index$SNP == rs, "Chr"]
			gwas_results[count, "POS"] <- probe_index[probe_index$SNP == rs, "Position"]
			gwas_results[count, "ID"] <- rs
			gwas_results[count, "REF"] <- "A"
			gwas_results[count, "ALT"] <- "T"
			gwas_results[count, "A1"] <- "T"
			gwas_results[count, "FIRTH?"] <- "NA"
			gwas_results[count, "TEST"] <- "ADD"
			gwas_results[count, "OBS_CT"] <- nrow(df)
			gwas_results[count, "OR"] <- coef(fit)[2]
			gwas_results[count, "LOG(OR)_SE"] <- summary(fit)$coeff[2,2]
			gwas_results[count, "L95"] <- CI[2,1]
			gwas_results[count, "U95"] <- CI[2,2]
			gwas_results[count, "Z_STAT"] <- summary(fit)$coeff[2,3]
			gwas_results[count, "P"] <- summary(fit)$coeff[2,4]
			gwas_results[count, "ERRCODE"] <- "."
			gwas_results[count, "PHENO"] <- p

		# Else, perform the logistic regression (forces usage of logistic instead of Firth regression)
		} else {
			fit <- glm(PHENO ~ . , data = df, family = binomial(link = "logit"))
			CI <- exp(confint.default(fit))

			# Fill in the result table following the PLINK scheme
			gwas_results[count, "#CHROM"] <- probe_index[probe_index$SNP == rs, "Chr"]
			gwas_results[count, "POS"] <- probe_index[probe_index$SNP == rs, "Position"]
			gwas_results[count, "ID"] <- rs
			gwas_results[count, "REF"] <- "A"
			gwas_results[count, "ALT"] <- "T"
			gwas_results[count, "A1"] <- "T"
			gwas_results[count, "FIRTH?"] <- "N"
			gwas_results[count, "TEST"] <- "ADD"
			gwas_results[count, "OBS_CT"] <- nrow(df)
			gwas_results[count, "OR"] <- exp(coef(fit)[2])
			gwas_results[count, "LOG(OR)_SE"] <- summary(fit)$coeff[2,2]
			gwas_results[count, "L95"] <- CI[2,1]
			gwas_results[count, "U95"] <- CI[2,2]
			gwas_results[count, "Z_STAT"] <- summary(fit)$coeff[2,3]
			gwas_results[count, "P"] <- summary(fit)$coeff[2,4]
			gwas_results[count, "ERRCODE"] <- "."
			gwas_results[count, "PHENO"] <- p

		}

		# Increase the counter 
		count <- count + 1
	}
}
rm(p, df_probes, df, fit, CI, count)

# Save all associations 
fwrite(gwas_results, paste0("/cauwerx/gwas/06_SCA/mirror/data/temp/associations/", sex, "/gwas_binary_mirror_SCA_", sex, "_round", round, ".txt.gz"), col.names = T, row.names = F, quote = F, sep = "\t")



#################################################
### STEP 6: identify significant associations ###

# Threshold for significance
thr <- 0.05/6633

# Subset significant associations
sig_gwas <- gwas_results[which(gwas_results$P <= thr), ]
print(paste0("Number of significant associations after round ", round, ": ", nrow(sig_gwas)))

# Save significant associations
fwrite(sig_gwas, paste0("/cauwerx/gwas/06_SCA/mirror/data/temp/significant_associations/", sex, "/sig_assoc_binary_mirror_SCA_", sex, "_round", round, ".txt.gz"), col.names = T, row.names = F, quote = F, sep = "\t", na = NA)



#################################################
### STEP 7: Update loop variables ###############

round <- round + 1
n <- nrow(sig_gwas)
rm(sig_gwas)
}
