# Select disease and model-specific probes (Fisher genotypic (2x3) test p-value <= 0.001; mirror freq/prune filter; >= 2 diseased carriers).

#################################################
### Libraries & External arguments ##############
library(data.table)
library(dplyr)
library(tidyr)

# Arguments
args <- commandArgs(trailingOnly = T)
sex <- args[1]
print(paste0("Analyzing: ", sex))

# Filter for probe selection
thr <- 0.001



#################################################
### STEP 1: Unique phenotypes ###################

# Retrieve all unique diseases (file header); File generated by "04_phenotypes/01_ICD10_extraction.R"
all_pheno <- names(fread(paste0("/cauwerx/gwas/04_phenotypes/data/pheno_ICD10_", sex, ".txt"), header = T, nrow = 0))[-1]

#################################################
### STEP 2: Select significant probes ###########

# Create an empty file to save the number of retained probes per phenotype and chromosme
summary <- data.frame(DISEASE = rep(all_pheno, each = 24), SEX = sex, CHR = rep(c(1:22, "X", "XY"), times = length(all_pheno)), PROBES_0.001 = NA, PROBES_MIRROR = NA, PROBES_DUP = NA, PROBES_DEL = NA)

# Loop over diseases
for (i in 1:nrow(summary)) { 

	# Define disease and chromosome
	p <- summary[i, "DISEASE"]
	chr <- summary[i, "CHR"]
	print(paste0("Probe selection for ", p, " - chr:", chr))

	# Load the file matching the phenotype and chromosome; Files generated by "05_gwas/03_probe_selection/01_get_probe_effects_*.sh"
	if (p == "DiseaseBurden") {
		df <- as.data.frame(fread(paste0("/cauwerx/gwas/05_gwas/03_probe_selection/data/temp/", sex, "/probe_effect_continuous_", sex, "_chr", chr, ".DiseaseBurden.glm.linear"), header = T))
	} else {
		df <- as.data.frame(fread(paste0("/cauwerx/gwas/05_gwas/03_probe_selection/data/temp/", sex, "/probe_effect_Fisher_", sex, "_chr", chr, ".", p, ".model"), header = T))
	}

	# If binary trait, retain only the p-values for the genotypic model
	if (p != "DiseaseBurden") {
		df <- df[which(df$TEST == "GENO"), ]
	} 

	# Retain significant probes and fill the summay table
	if (p == "DiseaseBurden") {
		df_sig <- as.data.frame(df[df$P <= thr & !is.na(df$P), "ID", drop = F])
	} else {
		df_sig <- as.data.frame(df[df$P <= thr & !is.na(df$P), names(df) %in% c("SNP", "A1", "A2", "AFF")])
		df_sig <- separate(df_sig, AFF, c("A1_AFF", "CN_AFF", "A2_AFF"), sep = "/", remove = T, convert = T)
	}
	summary[summary$DISEASE == p & summary$CHR == chr, "PROBES_0.001"] <- nrow(df_sig)
	print(paste0("Number of retained probes: ", nrow(df_sig)))

	# If no probe is significant or that we are dealing with the disease burden, save the file
	if(nrow(df_sig) == 0 | p == "DiseaseBurden") {

		# Mirror | U-shape
		fwrite(df_sig, paste0("/cauwerx/gwas/05_gwas/03_probe_selection/data/", sex, "/mirror/probes_mirror_", sex, "_chr", chr,".", p, ".txt"), row.names = F, col.names = F, sep = "\t", quote = F)
		summary[summary$DISEASE == p & summary$CHR == chr, "PROBES_MIRROR"] <- nrow(df_sig)
		# Duplication-only
		fwrite(df_sig, paste0("/cauwerx/gwas/05_gwas/03_probe_selection/data/", sex, "/duplication_only/probes_dup_", sex, "_chr", chr,".", p, ".txt"), row.names = F, col.names = F, sep = "\t", quote = F)
		summary[summary$DISEASE == p & summary$CHR == chr, "PROBES_DUP"] <- nrow(df_sig)
		# Deletion-only
		fwrite(df_sig, paste0("/cauwerx/gwas/05_gwas/03_probe_selection/data/", sex, "/deletion_only/probes_del_", sex, "_chr", chr,".", p, ".txt"), row.names = F, col.names = F, sep = "\t", quote = F)
		summary[summary$DISEASE == p & summary$CHR == chr, "PROBES_DEL"] <- nrow(df_sig)


	# Else, if the trait is binary, ensure that there are at least two affected individuals carriers of the relevant CNV type
	} else {

		# Mirror | U-shape
		df_sig_mirror <- df_sig[which(df_sig$A1_AFF + df_sig$A2_AFF >= 2), "SNP", drop = F]
		fwrite(df_sig_mirror, paste0("/cauwerx/gwas/05_gwas/03_probe_selection/data/", sex, "/mirror/probes_mirror_", sex, "_chr", chr,".", p, ".txt"), row.names = F, col.names = F, sep = "\t", quote = F)
		summary[summary$DISEASE == p & summary$CHR == chr, "PROBES_MIRROR"] <- nrow(df_sig_mirror)
		# Duplication-only
		df_sig$DUP <- df_sig$A1_AFF
		df_sig[which(df_sig$A2 == "T"), "DUP"] <- df_sig[which(df_sig$A2 == "T"), "A2_AFF"]
		df_sig_dup <- df_sig[which(df_sig$DUP >= 2), "SNP", drop = F]
		fwrite(df_sig_dup, paste0("/cauwerx/gwas/05_gwas/03_probe_selection/data/", sex, "/duplication_only/probes_dup_", sex, "_chr", chr,".", p, ".txt"), row.names = F, col.names = F, sep = "\t", quote = F)
		summary[summary$DISEASE == p & summary$CHR == chr, "PROBES_DUP"] <- nrow(df_sig_dup)
		# Deletion-only
		df_sig$DEL <- df_sig$A1_AFF
		df_sig[which(df_sig$A2 == "A"), "DEL"] <- df_sig[which(df_sig$A2 == "A"), "A2_AFF"]
		df_sig_del <- df_sig[which(df_sig$DEL >= 2), "SNP", drop = F]
		fwrite(df_sig_del, paste0("/cauwerx/gwas/05_gwas/03_probe_selection/data/", sex, "/deletion_only/probes_del_", sex, "_chr", chr,".", p, ".txt"), row.names = F, col.names = F, sep = "\t", quote = F)
		summary[summary$DISEASE == p & summary$CHR == chr, "PROBES_DEL"] <- nrow(df_sig_del)
	}

}



#################################################
### Save summary ################################

# Print summary per phenotype
print(aggregate(summary$PROBES_0.001, by = list(DISEASE = summary$DISEASE), FUN = sum))
fwrite(summary, paste0("/cauwerx/gwas/05_gwas/03_probe_selection/data/final/", sex, "/summary_probes_", sex, ".txt"), row.names = F, col.names = T, sep = "\t", quote = F)
