# Calculate the CNV region (CNVR) for each GWAS hit according to the U-shape model.

#################################################
### Libraries & External arguments ##############
library(data.table)
library(dplyr)
library(gtools)

# Arguments
args <- commandArgs(trailingOnly = T)
sex <- args[1]
print(paste0("Analyzing: ", sex))



#################################################
### STEP 1: Load data ###########################

# Load the last summary file from the SCA
files <- mixedsort(list.files(paste0("/home/cauwerx/scratch/cauwerx/projects/cnv_gwas/gwas/08_SCA/binary/white_british/u_shape/data/temp/summary/", sex), recursive = F, full.names = T))
df <- as.data.frame(fread(files[length(files)], header = T, select = c(3, 1, 2, 10:11, 15, 17)))
colnames(df)[2] <- "CHR"
print(paste0("Number of independent signals in ", sex,": ", nrow(df)))



#################################################
### STEP 2: Calculate the CNVR ##################

# Set parameters
r <- 0.5
dist <- 1500000

# Unique combination of Chr-rs
unique_combo <- df[!duplicated(df[, c(1,2)]), c(1,2,3)]
print(paste0("Number of probes to analyze: ", nrow(unique_combo)))

# Load all probes; This file contains the rs ID and position of all probes on the genotype array, available from the UKBB portal
probes <- as.data.frame(fread("/data/general_data/UKBB/probes.txt.gz", header = T, col.names = c("ID", "CHR", "POS")))
probes$CHR <- sub("23", "X", probes$CHR)

# Loop over each probes: identify adjacent probes
for(i in 1:nrow(unique_combo)) {

	# Define setting
	rs <- unique_combo[i, 1]
	chr <- unique_combo[i, 2]
	pos <- unique_combo[i, 3]
	print(paste0("Starting ", rs, " (chr", chr, ": ", pos, ") - number #", i))

	#### UPSTREAM REGION #####################################################

	# Modify the distance (2.5Mb) for problematic associations
	if (sex == "F" & rs == "rs2214831") {dist <- 2500000}
	if (sex == "All" & rs == "rs61752248") {dist <- 2500000}
	if (sex == "All" & rs == "rs77434182") {dist <- 2500000}

	# Save a temporary range file with rs number (capped at 0 bp)
	range_file <- paste0("/cauwerx/gwas/07_CNVR/u_shape/data/temp/", sex, "/", rs, "_up")
	if(pos-dist < 0) {fwrite(data.frame(CHR = chr, START = 1, END = pos, GROUP = "R"), range_file, col.names = F, row.names = F, quote = F, sep = "\t")
	} else {fwrite(data.frame(CHR = chr, START = pos-dist, END = pos, GROUP = "R"), range_file, col.names = F, row.names = F, quote = F, sep = "\t")}

	# Set the distance again 
	dist <- 1500000
	
	# Run PLINK: Extact probes within range
	input <- paste0("/data/UKBB/cnv_calls/PLINK/mirror/ukb_cnv_bed/ukb_cnv_chr", chr) # PLINK file set with mirror encoding; Generation of this file is described in Auwerx et al., 2022
	samples <- paste0("/cauwerx/gwas/01_samples/data/samples_white_british_plink1.9_", sex,".txt") # File generated by "01_samples/sample_filtering.R"
	system(paste("/cauwerx/softwares/plink/plink",
				 "--threads 20",	
				 "--bfile", input,
				 "--keep", samples,
				 "--extract range", range_file,
				 "--recode tab --out", range_file))

	# Load generated .ped and .map
	map <- as.data.frame(fread(paste0(range_file, ".map"), header = F, select = c(2,4), col.names = c("ID", "POS")))
	ped <- as.data.frame(fread(paste0(range_file, ".ped"), header = F, select = c(7:(6+nrow(map))), col.names = map$ID))

	# Correct the .ped in R and make all columns numerical
	ped[ped == "A A" | ped == "T T"] <- 1
	ped[ped == "A T" | ped == "T A"] <- 0
	for (j in 1:ncol(ped)) {ped[, j] <- as.numeric(ped[, j])}

	# Calculate the squared probe correlation matrix
	r2_matrix <- as.data.frame(cor(ped)^2)
	rm(ped)

	# Identify the first position in which r2 < r 
	map$cor <- r2_matrix[, which(colnames(r2_matrix) == rs)]
	map <- na.omit(map); rownames(map) <- 1:nrow(map)
	if(nrow(map) == nrow(map[which(map$cor > 0.5), ])) {row <- 1
	} else {row <- max(as.numeric(rownames(map[which(map$cor < r), ]))) + 1}  

	# Annotate table
	if(nrow(map) == row) {
		df[which(df$ID == rs), "START_CNVR"] <- map[row, "POS"]
		df[which(df$ID == rs), "NUM_PROBES_CNVR"] <- 1
	} else if (nrow(map)-row > 0) {
		df[which(df$ID == rs), "START_CNVR"] <- map[row, "POS"]
		df[which(df$ID == rs), "NUM_PROBES_CNVR"] <- nrow(map)-row+1
	}

	# Delete generated files
	unlink(range_file); unlink(paste0(range_file, ".log")); unlink(paste0(range_file, ".ped")); unlink(paste0(range_file, ".map"))
	rm(range_file)
 

	#### DOWNSTREAM REGION #####################################################

	# Save a temp range file with rs number
	range_file <- paste0("/cauwerx/gwas/07_CNVR/u_shape/data/temp/", sex, "/", rs, "_down")
	fwrite(data.frame(CHR = chr, START = pos, END = pos+dist, GROUP = "R"), range_file, col.names = F, row.names = F, quote = F, sep = "\t")
	
	# Run PLINK: Extact probes within range
	input <- paste0("/data/UKBB/cnv_calls/PLINK/mirror/ukb_cnv_bed/ukb_cnv_chr", chr) # PLINK file set with mirror encoding; Generation of this file is described in Auwerx et al., 2022
	samples <- paste0("/cauwerx/gwas/01_samples/data/samples_white_british_plink1.9_", sex,".txt") # File generated by "01_samples/sample_filtering.R"
	system(paste("/cauwerx/softwares/plink/plink",
				 "--threads 20",	
				 "--bfile", input,
				 "--keep", samples,
				 "--extract range", range_file,
				 "--recode tab --out", range_file))

	# Load generated .ped and .map
	map <- as.data.frame(fread(paste0(range_file, ".map"), header = F, select = c(2,4), col.names = c("ID", "POS")))
	ped <- as.data.frame(fread(paste0(range_file, ".ped"), header = F, select = c(7:(6+nrow(map))), col.names = map$ID))

	# Correct the .ped in R and make all columns numerical
	ped[ped == "A A" | ped == "T T"] <- 1
	ped[ped == "A T" | ped == "T A"] <- 0
	for (j in 1:ncol(ped)) {ped[, j] <- as.numeric(ped[, j])}

	# Calculate the squared probe correlation matrix
	r2_matrix <- as.data.frame(cor(ped)^2)
	rm(ped)

	# Identify the first position in which r2 < r 
	map$cor <- r2_matrix[, which(colnames(r2_matrix) == rs)]
	map <- na.omit(map); rownames(map) <- 1:nrow(map)
	if(nrow(map) == nrow(map[which(map$cor > 0.5), ])) {row <- nrow(map)
	} else {row <- min(as.numeric(rownames(map[which(map$cor < r), ]))) - 1}  

	# Annotate table
	if(row == 1) {
		df[which(df$ID == rs), "END_CNVR"] <- map[row, "POS"]
	} else if (row > 1) {
		df[which(df$ID == rs), "END_CNVR"] <- map[row, "POS"]
		df[which(df$ID == rs), "NUM_PROBES_CNVR"] <- df[which(df$ID == rs), "NUM_PROBES_CNVR"] + row - 1 # -1 to not count the lead probe double
	}

	# Delete generated files
	unlink(range_file); unlink(paste0(range_file, ".log")); unlink(paste0(range_file, ".ped")); unlink(paste0(range_file, ".map"))
	rm(range_file)


	#### GENERAL INFO ON REGION #################################################

	df[which(df$ID == rs), "LENGTH_CNVR"] <- df[which(df$ID == rs), "END_CNVR"] - df[which(df$ID == rs), "START_CNVR"]
	print(paste0("Length of associated CNVR: ", df[which(df$ID == rs), "LENGTH_CNVR"], " bp (", df[which(df$ID == rs), "NUM_PROBES_CNVR"], " probes)"))

}

# Reorder
df <- df[, c(1:8, 10, 11, 9)]

# Order by p-value
df <- df[order(df$P), ]

# Replace chromosome "XY" by "X"
df$CHR <- sub("XY", "X", df$CHR)



#################################################
### Save ########################################

# Print warning if some CNVR are were not fully captured due to the dist parameter
print(paste0("Warning?!"))
print(df[which((df$POS - df$START_CNVR) > 1400000 | (df$END_CNVR - df$POS) > 1400000), ])

# Save table
fwrite(df, paste0("/cauwerx/gwas/07_CNVR/u_shape/data/GWAS_CNVR_", sex,"_U.txt"), col.names = T, row.names = F, quote = F, sep = "\t")
