# Compare the Mb vs. Gene burden effects through simulation analyses (ICD10s).

#################################################
### Libraries & Paramteres ######################
library(data.table)
library(dplyr)

set.seed(8)
n <- 10000



#################################################
### STEP 1: Load data ###########################

# Genome-wide data
genome_size <- 3137161264 # UCSC GRCh37
hgnc_genes <- as.data.frame(fread("/cauwerx/softwares/ANNOVAR-20191024/humandb/hg19_refGene.txt"))

# Define traits to test (at least one significant burden association)
phenos <- c("IDA", "anaemia", "hypothyroidism", "T1D", "lipid", "SCZ", "bipolar", "apnea", "epilepsy", "cataract", "conduction", "HTN_essential", "IHD", "pneumonia", "asthma", "COPD", "OA", "osteoporosis", "AKI", "CKD") 

# Burden effects; Files generated by "09_burden_analysis/01_CNV_burden_binary.R"
burden_DEL <- as.data.frame(fread("/cauwerx/gwas/09_burden_analysis/data/no_CNVR_correction/DEL_burden_analysis_binary.txt", select = c(1,2,6,8)))
burden_DUP <- as.data.frame(fread("/cauwerx/gwas/09_burden_analysis/data/no_CNVR_correction/DUP_burden_analysis_binary.txt", select = c(1,2,6,8)))



#################################################
### STEP 2: Calculate GW gene density ###########

# Unique number of genes --> 28265 entities
u_genes <- unique(hgnc_genes$V13)

# Exclude MIRs --> 26289 entities
u_genes <- u_genes[grep("^MIR", u_genes, invert = T)]

# Calculate genome-wide gene density
gw_denisty <- length(u_genes)/(genome_size/1000000)
print(paste0("Estmiated GW gene density: ", round(gw_denisty, 2)))



#################################################
### STEP 3: Sig. gene density differences #######

# Create a dataframe to store the results
results <- data.frame(PHENO = rep(phenos, each = 2), TYPE = rep(c("DEL", "DUP"), times = length(phenos)))

# Loop over diseases with at least one significant association
for (p in phenos) {

	# Define the disease and subset the data
	print(paste0("The analyzed phenotype is: ", p))
	df_DEL <- burden_DEL[which(burden_DEL$PHENO == p), ]
	df_DUP <- burden_DUP[which(burden_DUP$PHENO == p), ]

	# Simulations (from a Gaussian distribution)
	sim_mb_DEL <- rnorm(n, mean = df_DEL[which(df_DEL$BURDEN == "Mb"), "BETA"], sd = df_DEL[which(df_DEL$BURDEN == "Mb"), "SE"])
	sim_genes_DEL <- rnorm(n, mean = df_DEL[which(df_DEL$BURDEN == "GENES"), "BETA"], sd = df_DEL[which(df_DEL$BURDEN == "GENES"), "SE"])
	sim_mb_DUP <- rnorm(n, mean = df_DUP[which(df_DUP$BURDEN == "Mb"), "BETA"], sd = df_DUP[which(df_DUP$BURDEN == "Mb"), "SE"])
	sim_genes_DUP <- rnorm(n, mean = df_DUP[which(df_DUP$BURDEN == "GENES"), "BETA"], sd = df_DUP[which(df_DUP$BURDEN == "GENES"), "SE"])

	# Calculate simulated ratios of Mb-to-gene effects
	ratio_DEL <- sim_mb_DEL/sim_genes_DEL
	ratio_DUP <- sim_mb_DUP/sim_genes_DUP

	# Estimate the SD of the simulated ratios
	SD_DEL <- sd(ratio_DEL)
	print(paste0("SD for DEL analysis: ", SD_DEL))
	SD_DUP <- sd(ratio_DUP)
	print(paste0("SD for DUP analysis: ", SD_DUP))

	# Calculate a t statistic (~1 sample t-test, SD of the true ratio can be ignored)
	t_DEL <- (gw_denisty-(df_DEL[which(df_DEL$BURDEN == "Mb"), "BETA"]/df_DEL[which(df_DEL$BURDEN == "GENES"), "BETA"]))/SD_DEL
	print(paste0("t for DEL analysis: ", t_DEL))
	t_DUP <- (gw_denisty-(df_DUP[which(df_DUP$BURDEN == "Mb"), "BETA"]/df_DUP[which(df_DUP$BURDEN == "GENES"), "BETA"]))/SD_DUP
	print(paste0("t for DUP analysis: ", t_DUP))

	# Calculate an associated p-value (two-sided)
	p_DEL <- 2*pnorm(-abs(t_DEL), mean = 0, sd = 1)
	print(paste0("p-value for DEL analysis: ", p_DEL))
	p_DUP <- 2*pnorm(-abs(t_DUP), mean = 0, sd = 1)
	print(paste0("p-value for DUP analysis: ", p_DUP))

	# Fill in the results table
	results[which(results$PHENO == p & results$TYPE == "DEL"), "Mb"] <- df_DEL[which(df_DEL$BURDEN == "Mb"), "BETA"]
	results[which(results$PHENO == p & results$TYPE == "DEL"), "GENES"] <- df_DEL[which(df_DEL$BURDEN == "GENES"), "BETA"]
	results[which(results$PHENO == p & results$TYPE == "DEL"), "RATIO"] <- df_DEL[which(df_DEL$BURDEN == "Mb"), "BETA"]/df_DEL[which(df_DEL$BURDEN == "GENES"), "BETA"]
	results[which(results$PHENO == p & results$TYPE == "DEL"), "SD_sim"] <- SD_DEL
	results[which(results$PHENO == p & results$TYPE == "DEL"), "T"] <- t_DEL
	results[which(results$PHENO == p & results$TYPE == "DEL"), "P"] <- p_DEL

	results[which(results$PHENO == p & results$TYPE == "DUP"), "Mb"] <- df_DUP[which(df_DUP$BURDEN == "Mb"), "BETA"]
	results[which(results$PHENO == p & results$TYPE == "DUP"), "GENES"] <- df_DUP[which(df_DUP$BURDEN == "GENES"), "BETA"]
	results[which(results$PHENO == p & results$TYPE == "DUP"), "RATIO"] <- df_DUP[which(df_DUP$BURDEN == "Mb"), "BETA"]/df_DUP[which(df_DUP$BURDEN == "GENES"), "BETA"]
	results[which(results$PHENO == p & results$TYPE == "DUP"), "SD_sim"] <- SD_DUP
	results[which(results$PHENO == p & results$TYPE == "DUP"), "T"] <- t_DUP
	results[which(results$PHENO == p & results$TYPE == "DUP"), "P"] <- p_DUP

}

# Print summary
print(paste0("Diseases with strict significant effect (p < 0.05/(n_pheno+1) = ", round(0.05/(length(phenos)+1), 4), "):"))
print(results[which(results$P <= 0.05/(length(phenos)+1)), ])

print(paste0("Diseases with nominally significant effect:"))
print(results[which(results$P <= 0.05), ])



#################################################
### Save results ################################

fwrite(results, "/cauwerx/gwas/09_burden_analysis/data/Mb_gene_effect_comparison/Mb_gene_effect_comparison_ICD10s.txt", col.names = T, row.names = F, sep = "\t", quote = F)
