# Calculate the CNV region (CNVR) for each GWAS hit according to the duplication-only model.

#################################################
### Libraries & External arguments ##############
library(data.table)
library(dplyr)
library(gtools)

# Arguments
args <- commandArgs(trailingOnly = T)
sex <- args[1]
print(paste0("Analyzing: ", sex))



#################################################
### STEP 1: Load data ###########################

# Load the last summary file from the SCA; File generated by "06_SCA/duplication_only/SCA_binary_dup.R"
files <- mixedsort(list.files(paste0("/cauwerx/gwas/06_SCA/duplication_only/data/temp/summary/", sex,"/"), recursive = F, full.names = T))
df <- as.data.frame(fread(files[length(files)], header = T, select = c(3, 1, 2, 10:11, 15, 17)))
colnames(df)[2] <- "CHR"
print(paste0("Number of independent signals in ", sex,": ", nrow(df)))



#################################################
### STEP 2: Calculate the CNVR ##################

# Unique combination of Chr-rs
unique_combo <- df[!duplicated(df[, c(1,2)]), c(1,2)]
print(paste0("Number of probes to analyze: ", nrow(unique_combo)))

# Load all probes; This file contains the rs ID and position of all probes on the genotype array, available from the UKBB portal
probes <- as.data.frame(fread("/data/general_data/UKBB/probes.txt.gz", header = T, col.names = c("ID", "CHR", "POS")))
probes$CHR <- sub("23", "X", probes$CHR)

# Loop over each probes: identify adjacent probes
for(i in 1:nrow(unique_combo)) {

	# Define setting
	rs <- unique_combo[i, 1]
	chr <- unique_combo[i, 2]
	print(paste0("Starting ", rs, " (chr", chr, ") - number #", i))

	# Save a temporary file with rs number
	rs_file <- paste0("/cauwerx/gwas/07_CNVR/duplication_only/data/temp/", sex, "/", rs)
	fwrite(data.frame(ID = rs), rs_file, col.names = F, row.names = F, quote = F, sep = "\t")
	
	# Run PLINK: detect probes in LD
	input <- paste0("/data/UKBB/cnv_calls/PLINK/duplication_only/ukb_cnv_bed/ukb_cnv_chr", chr) # PLINK file set with duplication encoding; Generation of this file is described in Auwerx et al., 2022
	samples <- paste0("/cauwerx/gwas/01_samples/data/samples_white_british_plink1.9_", sex,".txt") # File generated by "01_samples/sample_filtering.R"
	system(paste("/cauwerx/softwares/plink/plink",
				 "--fam /data/UKBB/cnv_calls/PLINK/mirror/ukb_cnv_bed/ukb_cnv_chrX_onlyF.fam", # Fake .fam file (all individuals = females) to deal with PLINK's male hemizygozity assumption (see Supplemental Note 1) 
				 "--threads 20",	
				 "--bfile", input,
				 "--keep", samples,
				 "--show-tags", rs_file,
				 "--tag-kb 3000 --tag-r2 0.5",
				 "--out", rs_file))
 	unlink(rs_file); unlink(paste0(rs_file, ".log"))

	# Detect the boundaries of the CNVR
	cnvr <- as.data.frame(fread(paste0(rs_file, ".tags"), header = F, col.names = "ID"))
	cnvr <- left_join(cnvr, probes, by = "ID")
	df[which(df$ID == rs), "START_CNVR"] <- min(cnvr$POS)
	df[which(df$ID == rs), "END_CNVR"] <- max(cnvr$POS)
	df[which(df$ID == rs), "LENGTH_CNVR"] <- max(cnvr$POS)-min(cnvr$POS)
	df[which(df$ID == rs), "NUM_PROBES_CNVR"] <- nrow(cnvr)
	print(paste0("Length of associated CNVR: ", max(cnvr$POS)-min(cnvr$POS), " kb (", nrow(cnvr), " probes)"))

}

# Order by p-value
df <- df[order(df$P), ]

# Replace chromosome "XY" by "X"
df$CHR <- sub("XY", "X", df$CHR)



#################################################
### Save ########################################

fwrite(df, paste0("/cauwerx/gwas/07_CNVR/duplication_only/data/GWAS_CNVR_", sex,"_dup.txt"), col.names = T, row.names = F, quote = F, sep = "\t")
