# Verify if associations are robust using an alternative modelling strategy (residual analysis).

#################################################
### Libraries & parameters ######################
library(data.table)
library(dplyr)

# Define the types of residuals
res <- "response"
print(paste0("Type of residuals for logistic regression: ", res))



#################################################
### STEP 1: Load data ###########################

# Load all CNV regions to analyze; File generated by "07_CNVR/combined/01_combine_models.R"
cnvrs <- as.data.frame(fread("/cauwerx/gwas/07_CNVR/combined/data/combined_CNVR/all_combined_CNVR.txt", header = F,
							 col.names = c("PHENO", "TOP_SNP", "CHR", "POS", "OR", "SE", "P", "START_CNVR", "END_CNVR", "MAIN_MODEL", "ALL_MODELS", "SIGN_MIRROR")))

# Load all probes; This file contains the rs ID and position of all probes on the genotype array, available from the UKBB portal
probes <- as.data.frame(fread("/data/general_data/UKBB/probes.txt.gz"))



#################################################
### STEP 2: Loop over signals ###################

# Create an empty dataframe to store data
results <- cnvrs
results$MODEL_alt <- NA
results$beta_alt <- NA
results$SE_alt <- NA
results$P_alt <- NA

# Loop over associations
for (i in 1:nrow(cnvrs)) {

	#################################################
	### STEP 2.1: Define the CNV region #############

	p <- cnvrs[i, "PHENO"]
	rs <- cnvrs[i, "TOP_SNP"]
	chr <- cnvrs[i, "CHR"]
	sex <- "All"; if(p %in% c("BC", "OC", "menstruation", "endometriosis")) {sex <- "F"}; if(p %in% c("PC")) {sex <- "M"}
	models <- unlist(strsplit(cnvrs[i, "ALL_MODELS"], "-"))
	print(paste0("Analyzing signal #", i,": ", p, " & ", rs, " (chr", chr,  ")", " --> models: ", models, "; population: ", sex))

	#################################################
	### STEP 2.2: Prepare the phenotype #############

	# Load slected phenotype; File generated by "04_phenotypes/01_ICD10_extraction.R"
	df_pheno <- as.data.frame(fread(paste0("/cauwerx/gwas/04_phenotypes/data/pheno_ICD10_", sex, ".txt")))
	df_pheno <- df_pheno[ , names(df_pheno) %in% c("IID", p)]
	colnames(df_pheno)[2] <- "PHENO" 
	if(p == "DiseaseBurden") {df_pheno$PHENO <- as.numeric(df_pheno$PHENO)
		} else {df_pheno$PHENO <- factor(df_pheno$PHENO, levels = c(1,2), labels = c(0,1))}

	# Load selected covariates
	df_cov <- as.data.frame(fread("/cauwerx/gwas/03_covariates/data/covariates.txt")) # File generated by "03_covariates/01_covariate_extraction.R"
	if (file.size(paste0("/cauwerx/gwas/05_gwas/02_covariate_selection/data/", sex,"/covariates_", sex,".", p, ".txt")) > 1) { # File generated by "05_gwas/02_covariate_selection/01_covariate_selection.R"
		cov_to_add <- as.data.frame(fread(paste0("/cauwerx/gwas/05_gwas/02_covariate_selection/data/", sex,"/covariates_", sex,".", p, ".txt"), header = F, col.names = "covariates",  sep = "\t"))
		cov_to_add <- unlist(strsplit(cov_to_add[1, "covariates"], ","))
		df_cov <- df_cov[, names(df_cov) %in% c("IID", cov_to_add)]
	} else {
		df_cov <- df_cov[, "IID", drop = F]
	}

	# Merge disease/covariates into a single dataframe
	df_pheno <- na.omit(left_join(df_pheno, df_cov, by = "IID"))
	rm(df_cov)

	# Logistic or linear regression to correct the binary disease status
	if(p == "DiseaseBurden") {
		df_pheno$PHENOc <- residuals(lm(PHENO ~ . , data = df_pheno[-1]), type = res)
	} else {
		df_pheno$PHENOc <- residuals(glm(PHENO ~ . , data = df_pheno[-1], family = binomial(link = "logit")), type = res)
	}

	# Inverse normal transformation
	df_pheno$PHENOc_INT <- qnorm((rank(df_pheno$PHENOc, na.last = "keep")-0.5)/sum(!is.na(df_pheno$PHENOc)))
	 

	
	#################################################
	### STEP 2.3: Prepare the genotype ##############

	# If variant is on chromosome X, verify if it is on the PAR
	if (chr == "X") { if(probes[which(probes$SNP == rs), "Chr"] == "23")  {chr <- "XY"}	}

	# Loop over the models to extract the CNV profiles
	for (m in models) {

		# Select the appropriate PLINK file
		if(m == "U" | m == "M") {
			input <- paste0("/data/UKBB/cnv_calls/PLINK/mirror/ukb_cnv_bed/ukb_cnv_chr", chr)} # PLINK file set with mirror encoding; Generation of this file is described in Auwerx et al., 2022
		if(m == "DUP") {
			input <- paste0("/data/UKBB/cnv_calls/PLINK/duplication_only/ukb_cnv_bed/ukb_cnv_chr", chr)} # PLINK file set with duplication encoding; Generation of this file is described in Auwerx et al., 2022
		if(m == "DEL") {
			input <- paste0("/data/UKBB/cnv_calls/PLINK/deletion_only/ukb_cnv_bed/ukb_cnv_chr", chr)} # PLINK file set with deletion encoding; Generation of this file is described in Auwerx et al., 2022

		# Generate an input/output file
		output <- paste0("/cauwerx/gwas/07_CNVR/combined/data/temp/cnv_profiles/",p, "_", m, "_", rs)
		fwrite(data.frame(rs), output, col.names = F, row.names = F, quote = F, sep = "\t")

		# Run PLINK v1.9 to extract the variant
		samples <- paste0("/cauwerx/gwas/01_samples/data/samples_white_british_plink1.9_", sex, ".txt") # File generated by "01_samples/sample_filtering.R"
		system(paste("/cauwerx/softwares/plink/plink",
					 "--bfile", input, 
					 "--fam /data/UKBB/cnv_calls/PLINK/mirror/ukb_cnv_bed/ukb_cnv_chrX_onlyF.fam", # Fake .fam file (all individuals = females) to deal with PLINK's male hemizygozity assumption (see Supplemental Note 1) 
					 "--keep", samples,
					 "--extract", output,
					 "--recode tab --out", output,
					 "--threads 20")) 
	}


	#################################################
	### STEP 2.4: Association #######################

	# Generate a temporary dataframe to store the association results
	df_temp <- data.frame()

	# Loop over the models to perform the association study
	for (m in models) {

		# Correct the .ped in R
		output <- paste0("/cauwerx/gwas/07_CNVR/combined/data/temp/cnv_profiles/",p, "_", m, "_", rs)
		df_geno <- as.data.frame(fread(paste0(output, ".ped"), header = F, select = c(2,7), col.names = c("IID", "CNV")))	

		if(m == "M") {
			df_geno[which(df_geno$CNV == "A A"), "CNV"] <- -1
			df_geno[which(df_geno$CNV == "A T" | df_geno$CNV == "T A"), "CNV"] <- 0
			df_geno[which(df_geno$CNV == "T T"), "CNV"] <- 1
			df_geno[which(df_geno$CNV == "0 0"), "CNV"] <- NA}
		if(m == "U") {
			df_geno[which(df_geno$CNV == "A A" | df_geno$CNV == "T T"), "CNV"] <- 1
			df_geno[which(df_geno$CNV == "A T" | df_geno$CNV == "T A"), "CNV"] <- 0
			df_geno[which(df_geno$CNV == "0 0"), "CNV"] <- NA}
		if(m == "DUP" | m == "DEL") {
			df_geno[which(df_geno$CNV == "T T"), "CNV"] <- 1
			df_geno[which(df_geno$CNV == "A T" | df_geno$CNV == "T A"), "CNV"] <- 0
			df_geno[which(df_geno$CNV == "0 0"), "CNV"] <- NA}

		df_geno$CNV <- as.numeric(df_geno$CNV)

		# Clean
		unlink(output); unlink(paste0(output, ".log")); unlink(paste0(output, ".map")); unlink(paste0(output, ".ped"))
		rm(input, output, samples)

		# Merge corrected phenotype and genotype data
		df <- left_join(df_pheno[, c("IID", "PHENOc_INT")], df_geno, by = c("IID"))

		# Phenotype vs. CNV
		fit <- lm(PHENOc_INT ~ CNV, data = df)

		# Store results
		df_temp_temp  <- data.frame(MODEL_alt = m, beta_alt = summary(fit)$coeff[2, 1], SE_alt = summary(fit)$coeff[2, 2], P_alt = summary(fit)$coeff[2, 4])
		df_temp <- rbind(df_temp, df_temp_temp)

	}



	#################################################
	### STEP 2.5: Select the most significant signal 

	# Select the model yielding the smallest p-value
	top_m  <- which.min(df_temp$P_alt)

	# Add results
	results[i, "MODEL_alt"] <- as.character(df_temp[top_m, "MODEL_alt"])
	results[i, "beta_alt"] <- df_temp[top_m, "beta_alt"]
	results[i, "SE_alt"] <- df_temp[top_m, "SE_alt"]
	results[i, "P_alt"] <- df_temp[top_m, "P_alt"]
	rm(df_temp, df_temp_temp)

}



#################################################
### STEP 3: Identify significant signals ########

results_sig <- results[which(results$P_alt <= 7.5e-6), ]
print(paste0("Number of genome-wide signals: ", nrow(results_sig), "/", nrow(results)))



#################################################
### Save ########################################

fwrite(results, paste0("/cauwerx/gwas/07_CNVR/combined/data/residual_analysis/all_residual_analysis_", res,".txt"), col.names = T, row.names = F, sep = "\t", quote = F)
fwrite(results_sig, paste0("/cauwerx/gwas/07_CNVR/combined/data/residual_analysis/significant_residual_analysis_", res, ".txt"), col.names = T, row.names = F, sep = "\t", quote = F)
