# Verify if associations are robust using an alternative modelling strategy (Cox proportional-hazards model).

#################################################
### Libraries ###################################
library(data.table)
library(dplyr)
library(tidyr)
library(survival)



#################################################
### STEP 1: Load data ###########################

# All CNV regions to analyze (remove those with the dissease burden); File generated by "07_CNVR/combined/01_combine_models.R"
cnvrs <- as.data.frame(fread("/cauwerx/gwas/07_CNVR/combined/data/combined_CNVR/all_combined_CNVR.txt", header = F, col.names = c("PHENO", "TOP_SNP", "CHR", "POS", "OR", "SE", "P",  "START_CNVR", "END_CNVR", "MAIN_MODEL", "ALL_MODELS", "SIGN_MIRROR")))
cnvrs <- cnvrs[!cnvrs$PHENO %in% c("DiseaseBurden"), ]
print(paste0("Number of associations to analyze: ", nrow(cnvrs)))

# Load covariates; File generated by "03_covariates/01_covariate_extraction.R"
cov <- as.data.frame(fread("/cauwerx/gwas/03_covariates/data/covariates.txt", header = T))

# Load all probes; This file contains the rs ID and position of all probes on the genotype array, available from the UKBB portal
probes <- as.data.frame(fread("/data/general_data/UKBB/probes.txt.gz"))



#################################################
### STEP 2: Loop over CNVR associations #########

# Ceate empty dataframes to store results
coxPH_fit_results <- data.frame()
coxPH_top_fit_results <- data.frame()

# Loop over associations
for (i in 1:nrow(cnvrs)) {

	# Define the association
	pheno <- as.character(cnvrs[i, "PHENO"])
	rs <- as.character(cnvrs[i, "TOP_SNP"])
	chr <- as.character(cnvrs[i, "CHR"]); if (chr == "X") {if(probes[which(probes$SNP == rs), "Chr"] == "23") {chr <- "XY"}}
	pos <- cnvrs[i, "POS"]
	or <- cnvrs[i, "OR"]
	p <- cnvrs[i, "P"]
	main_model <- as.character(cnvrs[i, "MAIN_MODEL"])
	start <- cnvrs[i, "START_CNVR"]
	end <- cnvrs[i, "END_CNVR"]
	sex <- "All"; if (pheno %in% "PC") {sex <- "M"}; if (pheno %in% c("BC", "OC", "menstruation", "endometriosis")) {sex <- "F"}
	print(paste0("Analyzing association #", i, ": ", rs, " (chr", chr, ":", pos, ") & ", pheno))


	##########################################################
	# Part I: Load time-to-event data

	# Load age at diagnosis data; File generated by "04_phenotypes/03_time_to_event.R"
	coxPH <- as.data.frame(fread(paste0("/cauwerx/gwas/04_phenotypes/data/time_to_event/time_to_event_", sex, ".", pheno, ".txt.gz"), header = T))


	##########################################################
	# Part II: Load phenotype-specific covariates
	# Note: Check that there is at least one retained covariate

	if (file.size(paste0("/cauwerx/gwas/05_gwas/02_covariate_selection/data/", sex, "/covariates_", sex, ".", pheno, ".txt")) > 1) { # File generated by "05_gwas/02_covariate_selection/01_covariate_selection.R"
		selected_cov <- unlist(strsplit(as.data.frame(fread(paste0("/cauwerx/gwas/05_gwas/02_covariate_selection/data/", sex, "/covariates_", sex, ".", pheno, ".txt"), header = F, sep = "\t"))[1,1], ","))
		selected_cov <- cov[, names(cov) %in% c("IID", selected_cov)]
		coxPH <- left_join(coxPH, selected_cov, by = "IID")
	} 
	rm(selected_cov)


	##########################################################
	# Part III: Extract the CNV profile 

	# Write a temporary file
	rs_file <- paste0("/cauwerx/gwas/07_CNVR/binary/white_british/combined/data/temp/coxPH/", rs)
	fwrite(data.frame(rs), rs_file, col.names = F, row.names = F, quote = F, sep = "\t")

	# Extract profile with PLINK
	input <- paste0("/data/UKBB/cnv_calls/PLINK/mirror/ukb_cnv_bed/ukb_cnv_chr", chr) # PLINK file set with mirror encoding; Generation of this file is described in Auwerx et al., 2022
	sample_file <- paste0("/cauwerx/gwas/01_samples/data/samples_white_british_plink1.9_", sex, ".txt") # File generated by "01_samples/sample_filtering.R"
	system(paste("/home/cauwerx/scratch/cauwerx/softwares/plink/plink",
				 "--fam /data/UKBB/cnv_calls/PLINK/mirror/ukb_cnv_bed/ukb_cnv_chrX_onlyF.fam", # Fake .fam file (all individuals = females) to deal with PLINK's male hemizygozity assumption (see Supplemental Note 1)
				 "--threads 20",
				 "--bfile", input, 
				 "--keep", sample_file,
				 "--extract", rs_file,
				 "--recode tab --out", rs_file))

	# Encode the .ped following a mirror model & delete the rs_file and the .ped/.map
	geno <- as.data.frame(fread(paste0(rs_file, ".ped"), header = F, select = c(2,7), col.names = c("IID", "MIRROR")))
	geno[which(geno$MIRROR == "A A"), "MIRROR"] <- -1
	geno[which(geno$MIRROR == "A T" | geno$MIRROR == "T A"), "MIRROR"] <- 0
	geno[which(geno$MIRROR == "T T"), "MIRROR"] <- 1
	unlink(rs_file); unlink(paste0(rs_file, ".log")); unlink(paste0(rs_file, ".map")); unlink(paste0(rs_file, ".ped"))

	# Add a U-shape encoding
	geno$U_SHAPE <- geno$MIRROR
	geno[which(geno$MIRROR == -1), "U_SHAPE"] <- 1

	# Add a duplication-only encoding
	geno$DUP<- geno$MIRROR
	geno[which(geno$MIRROR == -1), "DUP"] <- NA

	# Add a deletion-only encoding
	geno$DEL <- geno$MIRROR
	geno[which(geno$MIRROR == -1), "DEL"] <- 1
	geno[which(geno$MIRROR == 1), "DEL"] <- NA


	##########################################################
	# Part IV: CoxPH analysis accordig to the 4 association models

	# Create a temporary dataframe to store the results of the 4 models
	df_temp <- data.frame(PHENO = rep(pheno, 4), 
						  TOP_SNP = rep(rs, 4), CHR = rep(chr, 4), POS = rep(pos, 4), START = rep(start, 4), END = rep(end, 4),
						  OR_GWAS = rep(or, 4), P_GWAS = rep(p, 4), MODEL_GWAS = rep(main_model, 4), 
						  MODEL_coxPH = c("M", "DUP", "DEL", "U"),COEF_coxPH = NA, HR_coxPH = NA, SE_coxPH = NA, P_coxPH = NA)


	# Mirror
	df <- na.omit(left_join(coxPH, geno[, names(geno) %in% c("IID", "MIRROR")], by = "IID"))
	df$MIRROR <- as.numeric(df$MIRROR)
	if ("sex" %in% names(df)) {df$sex <- factor(df$sex, levels = c(1,2), labels = c("M", "F"))}
	if ("array" %in% names(df)) {df$array <- factor(df$array, levels = c("UKBB","UKBL"))}
	fit_mirror <- coxph(Surv(time = ALHM, event = INDICATOR) ~ ., data = df[, -1])
	df_temp[1, "COEF_coxPH"] <- summary(fit_mirror)$coefficients["MIRROR", 1]
	df_temp[1, "HR_coxPH"] <- summary(fit_mirror)$coefficients["MIRROR", 2]
	df_temp[1, "SE_coxPH"] <- summary(fit_mirror)$coefficients["MIRROR", 3]
	df_temp[1, "P_coxPH"] <- summary(fit_mirror)$coefficients["MIRROR", 5]
	rm(df, fit_mirror)

	# Duplication-only
	df <- na.omit(left_join(coxPH, geno[, names(geno) %in% c("IID", "DUP")], by = "IID"))
	df$DUP <- as.numeric(df$DUP)
	if ("sex" %in% names(df)) {df$sex <- factor(df$sex, levels = c(1,2), labels = c("M", "F"))}
	if ("array" %in% names(df)) {df$array <- factor(df$array, levels = c("UKBB","UKBL"))}
	fit_dup <- coxph(Surv(time = ALHM, event = INDICATOR) ~ ., data = df[, -1])
	df_temp[2, "COEF_coxPH"] <- summary(fit_dup)$coefficients["DUP", 1]
	df_temp[2, "HR_coxPH"] <- summary(fit_dup)$coefficients["DUP", 2]
	df_temp[2, "SE_coxPH"] <- summary(fit_dup)$coefficients["DUP", 3]
	df_temp[2, "P_coxPH"] <- summary(fit_dup)$coefficients["DUP", 5]
	rm(df, fit_dup)

	# Deletion-only
	df <- na.omit(left_join(coxPH, geno[, names(geno) %in% c("IID", "DEL")], by = "IID"))
	df$DEL <- as.numeric(df$DEL)
	if ("sex" %in% names(df)) {df$sex <- factor(df$sex, levels = c(1,2), labels = c("M", "F"))}
	if ("array" %in% names(df)) {df$array <- factor(df$array, levels = c("UKBB","UKBL"))}
	fit_del <- coxph(Surv(time = ALHM, event = INDICATOR) ~ ., data = df[, -1])
	df_temp[3, "COEF_coxPH"] <- summary(fit_del)$coefficients["DEL", 1]
	df_temp[3, "HR_coxPH"] <- summary(fit_del)$coefficients["DEL", 2]
	df_temp[3, "SE_coxPH"] <- summary(fit_del)$coefficients["DEL", 3]
	df_temp[3, "P_coxPH"] <- summary(fit_del)$coefficients["DEL", 5]
	rm(df, fit_del)

	# U-shape
	df <- na.omit(left_join(coxPH, geno[, names(geno) %in% c("IID", "U_SHAPE")], by = "IID"))
	df$U_SHAPE <- as.numeric(df$U_SHAPE)
	if ("sex" %in% names(df)) {df$sex <- factor(df$sex, levels = c(1,2), labels = c("M", "F"))}
	if ("array" %in% names(df)) {df$array <- factor(df$array, levels = c("UKBB","UKBL"))}
	fit_U <- coxph(Surv(time = ALHM, event = INDICATOR) ~ ., data = df[, -1])
	df_temp[4, "COEF_coxPH"] <- summary(fit_U)$coefficients["U_SHAPE", 1]
	df_temp[4, "HR_coxPH"] <- summary(fit_U)$coefficients["U_SHAPE", 2]
	df_temp[4, "SE_coxPH"] <- summary(fit_U)$coefficients["U_SHAPE", 3]
	df_temp[4, "P_coxPH"] <- summary(fit_U)$coefficients["U_SHAPE", 5]
	rm(df, fit_U)


	##########################################################
	# Part V: Integrate results

	df_temp <- na.omit(df_temp)
	coxPH_fit_results <- rbind(coxPH_fit_results, df_temp)
	coxPH_top_fit_results <- rbind(coxPH_top_fit_results, df_temp[which.min(df_temp$P_coxPH), ])

}

print(paste0("Number of associations validated by coxPH (p < 7.5e-6): ", nrow(coxPH_top_fit_results[which(coxPH_top_fit_results$P_coxPH < 7.5e-6), ]), "/", nrow(cnvrs)))
print(paste0("Number of associations with more significant p-value with the coxPH modelling strategy: ", nrow(coxPH_top_fit_results[which(coxPH_top_fit_results$P_coxPH < coxPH_top_fit_results$P_GWAS), ]), "/", nrow(cnvrs)))



#################################################
### Save results ################################

fwrite(coxPH_fit_results, "/cauwerx/gwas/07_CNVR/combined/data/coxPH/coxPH_fit_results.txt", col.names = T, row.names = T, quote = F, sep = "\t")
fwrite(coxPH_top_fit_results, "/cauwerx/gwas/07_CNVR/combined/data/coxPH/coxPH_top_fit_results.txt", col.names = T, row.names = F, quote = F, sep = "\t")
