# Identify unique, non-overlapping CNV regions

#################################################
### Libraries ###################################
library(data.table)
library(dplyr)



#################################################
### STEP 1: Load CNVRs ###########################

# All CNV regions to analyze (remove those with the disease burden); File generated by "07_CNVR/combined/01_combine_models.R"
cnvrs <- as.data.frame(fread("/cauwerx/gwas/07_CNVR/combined/data/combined_CNVR/all_combined_CNVR.txt", header = F, select = c(1,3,8,9,11), col.names = c("PHENO", "CHR", "START_CNVR", "END_CNVR", "ALL_MODELS")))
cnvrs <- cnvrs[, c(2:5)] 

# Load all probes; This file contains the rs ID and position of all probes on the genotype array, available from the UKBB portal
probes <- as.data.frame(fread("/data/general_data/UKBB/probes.txt.gz", select = c(2,3), col.names = c("CHR", "POS")))
probes$CHR <- as.numeric(gsub("X", 23, probes$CHR))



#################################################
### STEP 2: Non-overlapping regions #############

# Make the CHR column numeric & order regions accordingly
cnvrs$CHR <- as.numeric(gsub("X", 23, cnvrs$CHR))
cnvrs <- cnvrs[order(cnvrs$CHR, cnvrs$START_CNVR, decreasing = F), ]

# Define unique CNVRs
u_cnvrs <- data.table(cnvrs[1,])
for (i in 2:nrow(cnvrs)){

	# Define the region
	test <- data.table(cnvrs[i,])
	setkey(test, CHR, START_CNVR, END_CNVR)
	match <- na.omit(foverlaps(u_cnvrs, test, type = "any", which = T))

	# Verify if there is an overlap
	if(nrow(match) == 0) {
		u_cnvrs <- rbind(u_cnvrs, test)
	} else {
		u_cnvrs[as.numeric(match[1,1]), "START_CNVR"] <- min(test$START_CNVR, as.numeric(u_cnvrs[as.numeric(match[1,1]), "START_CNVR"]))
		u_cnvrs[as.numeric(match[1,1]), "END_CNVR"] <- max(test$END_CNVR, as.numeric(u_cnvrs[as.numeric(match[1,1]), "END_CNVR"]))
		u_cnvrs[as.numeric(match[1,1]), "ALL_MODELS"] <- paste(unique(c(unlist(strsplit(test$ALL_MODELS, "-")), unlist(strsplit(as.character(u_cnvrs[as.numeric(match[1,1]), "ALL_MODELS"]), "-")))), collapse = "-")
	}

}; rm(i, test, match)



#################################################
### STEP 3: Annotate CNVRs ######################

### Length bp ###################################

u_cnvrs$Length_bp <- (u_cnvrs$END_CNVR - u_cnvrs$START_CNVR) + 1 


### Length probes ###############################

for(i in 1:nrow(u_cnvrs)) {
	u_cnvrs[i, "Length_probes"] <- nrow(probes[which(probes$CHR == as.numeric(u_cnvrs[i, "CHR"]) & probes$POS >= as.numeric(u_cnvrs[i, "START_CNVR"]) & probes$POS <= as.numeric(u_cnvrs[i, "END_CNVR"])), ])
} 


### Number of genes #############################

# Remodel & Save ANNOVAR input
cnvrs <- data.table(CHR = u_cnvrs$CHR, START = u_cnvrs$START_CNVR, END = u_cnvrs$END_CNVR,
					REF = 0, OBS = "-",
					ALL_MODELS = u_cnvrs$ALL_MODELS, Length_bp = u_cnvrs$Length_bp, Length_probes = u_cnvrs$Length_probes)
cnvrs$CHR <- gsub(23, "X", cnvrs$CHR)
annovar_input <- "/cauwerx/gwas/07_CNVR/combined/data/temp/ANNOVAR/ANNOVAR_input_uCNVRs.txt"
fwrite(cnvrs, annovar_input, col.names = F, row.names = F, quote = F, sep = "\t")

# Run ANNOVAR in terminal - HGNC
output_HGNC <- "/cauwerx/gwas/07_CNVR/combined/data/temp/ANNOVAR/CNVR_ANNOVAR_HGNC"
system(paste("/cauwerx/softwares/ANNOVAR-20191024/annotate_variation.pl",
			 "--geneanno -dbtype refGene",
			 "-out", output_HGNC, 
		     "-build hg19", 
			 annovar_input,
			 "/cauwerx/softwares/ANNOVAR-20191024/humandb/"))
unlink(paste0(output_HGNC, ".log"))

# Load the files
u_cnvrs <- as.data.frame(fread(paste0(output_HGNC, ".variant_function"), header = F, select = c(1:5, 8:10), col.names = c("REGION", "GENES", "CHR", "START_CNVR", "END_CNVR", "ALL_MODELS", "Length_bp", "Length_probes")))

# Count the number of genes (Note: only genes with protein-coding region affected are counted)
for (i in 1:nrow(u_cnvrs)) {
	u_cnvrs[i, "Length_genes"] <- length(unlist(strsplit(as.character(u_cnvrs[i, "GENES"]), ",")))
}
u_cnvrs[which(u_cnvrs$REGION != "exonic"), "Length_genes"] <- 0

### Gene density ##############################

u_cnvrs$GD_Mb <- u_cnvrs$Length_genes / (u_cnvrs$Length_bp/1000000)



#################################################
### STEP 4: Separate DUPs/DELs ##################

# Create two columns to determine to which group the region belongs
u_cnvrs$DUP <- 0
u_cnvrs$DEL <- 0

# Assign to both DUP & DEL regions tagged with both or neither
u_cnvrs[(grepl("DUP", u_cnvrs$ALL_MODELS) & grepl("DEL", u_cnvrs$ALL_MODELS)), c(11,12)] <- 1
u_cnvrs[(!grepl("DUP", u_cnvrs$ALL_MODELS) & !grepl("DEL", u_cnvrs$ALL_MODELS)), c(11,12)] <- 1

# Assign to DUP regions tagged with DUP but not DEL
u_cnvrs[(grepl("DUP", u_cnvrs$ALL_MODELS) & !grepl("DEL", u_cnvrs$ALL_MODELS)), "DUP"] <- 1

# Assign to DEL regions tagged with DUP but not DUP
u_cnvrs[(grepl("DEL", u_cnvrs$ALL_MODELS) & !grepl("DUP", u_cnvrs$ALL_MODELS)), "DEL"] <- 1


#################################################
### STEP 5: Save ################################

fwrite(u_cnvrs, "/cauwerx/gwas/07_CNVR/combined/data/unique_CNVR/unique_CNVRs_annotated.txt", col.names = T, row.names = F, sep = "\t", quote = F)
