# Filter probes for a CNV frequency > 0.01% and prune at r 0.9999

#################################################
### Libraries & External arguments ##############
library(data.table)
library(dplyr)
library(tidyr)
library(stringr)

# Arguments
args <- commandArgs(trailingOnly = TRUE)
chr <- args[1]
print(paste0("Analyzing chromosome ", chr))



#################################################
### STEP 1: Load data & Filter ##################

# Load all genotype frequencies; Generation of this file is described in Auwerx et al., 2022 (script "GWAS/02_probes/frequency/03_genotype_frequency_All.R")
geno_freq <- as.data.frame(fread("/data/UKBB/cnv_calls/genotype_frequency/genotype_freq_WB_All.txt.gz", header = T))

# Filter for CNV frequency (>= 0.01%)
geno_freq <- geno_freq[which(geno_freq$Chr == chr & geno_freq$FreqCNV >= 0.01), ]
print(paste0("Number of probes with CNV frequency >= 0.01% on chromosome ", chr, ": ", nrow(geno_freq)))

# Save temporarily for pruning
freq_filter_file <- paste0("/cauwerx/gwas/02_probes/data/temp/freq_filter/probes_WB_All_CNV_0.01_chr", chr, ".txt")
fwrite(geno_freq[, 2, drop = F], freq_filter_file, col.names = F, row.names = F, quote = F, sep = "\t")



#################################################
### STEP 2: Pruning with PLINK v2 ###############

# Define the pruning threshold
thr_r <- 0.9999

# Define input/output files
input <- paste0("/data/UKBB/cnv_calls/PLINK/mirror/ukb_cnv_bed/ukb_cnv_chr", chr) # PLINK file set with mirror encoding; Generation of this file is described in Auwerx et al., 2022 
output <- paste0("/cauwerx/gwas/02_probes/data/pruning/probes_WB_All_prune_0.9999_CNV_chr", chr)

# Run PLINK v2
system(paste("/cauwerx/softwares/plink2/plink2",
			 "--bfile", input,
			 "--keep /cauwerx/gwas/01_samples/data/samples_white_british_All.txt", # File generated by "01_samples/sample_filtering.R"
	 		 "--extract", freq_filter_file,
			 "--indep-pairwise 500 250", thr_r,				 
 			 "--out", output))

# Delete the log and temporary files file
unlink(paste0(output, ".log"))
unlink(freq_filter_file)

# Assess the PLINK run
print(paste0("Number of remaining probes on chromosome ", chr, " after frequency (all CNV) selection and pruning at r ", thr_r, ": ", nrow(as.data.frame(fread(paste0(output, ".prune.in"), header = F)))))
