# Identify CNVRs harboring OMIM morbid genes.

#################################################
### Libraries ###################################
library(data.table)
library(dplyr)



#################################################
### STEP 1: Load data ###########################

# OMIM morbid genes; Downloaded from the OMIM database (07/2022)
morbid <- as.data.frame(fread("/data/OMIM/OMIM_morbidmap_220727.txt", skip = 4, sep = "\t", fill = T, select = 1:3, col.names = c("DISEASE", "HGNC", "OMIM")))

# OMIM gene synonyms; Downloaded from the OMIM database (07/2022)
genes_syn <- as.data.frame(fread("/data/OMIM/OMIM_genemap2_220727.txt", skip = 4, sep = "\t", fill = T, select = c(6,11), col.names = c("OMIM", "ENSG")))

# CNVR genes - HGNC; File generated by "08_ANNOVAR/01_gene_annotation_ANNOVAR.R"
cnvrs_hgnc <- as.data.frame(fread("/cauwerx/gwas/08_ANNOVAR/combined/data/gene_annotation/CNVR_ANNOVAR_HGNC.variant_function", header = F, select = c(8,3,4,5,2), col.names = c("PHENO", "CHR", "START_CNVR", "END_CNVR", "HGNC")))

# CNVR genes - ENSG; File generated by "08_ANNOVAR/01_gene_annotation_ANNOVAR.R"
cnvrs_ensg <- as.data.frame(fread("/cauwerx/gwas/08_ANNOVAR/combined/data/gene_annotation/CNVR_ANNOVAR_ENSEMBL.variant_function", header = F, select = c(8,3,4,5,2), col.names = c("PHENO", "CHR", "START_CNVR", "END_CNVR", "ENSG")))



#################################################
### STEP 2: Modify OMIM data ####################

# Add ENSG 
morbid <- left_join(morbid, genes_syn, by = "OMIM")

# Remove synonym gene names & transcript names
morbid$HGNC <- gsub(",.*", "", morbid$HGNC)
morbid$ENSG <- gsub(",.*", "", morbid$ENSG)



#################################################
### STEP 3: Annotate CNVRs with morbid genes ####

# Create an empty dataframe to store the results
cnv_morbid <- data.frame()

# Loop over the CNVRs
for (i in 1:nrow(cnvrs_hgnc)) {

	# Identify genes overlapping the CNV region - HGNC
	genes_hgnc <- unlist(strsplit(cnvrs_hgnc[i, "HGNC"], split = ","))
	genes_hgnc <- data.frame(PHENO = rep(cnvrs_hgnc[i, "PHENO"], length(genes_hgnc)),
							 CHR = rep(cnvrs_hgnc[i, "CHR"], length(genes_hgnc)),
							 START_CNVR = rep(cnvrs_hgnc[i, "START_CNVR"], length(genes_hgnc)),
							 END_CNVR = rep(cnvrs_hgnc[i, "END_CNVR"], length(genes_hgnc)),
							 HGNC = genes_hgnc, stringsAsFactors = F)

	# Identify genes overlapping the CNV region - ENSG
	genes_ensg <- unlist(strsplit(cnvrs_ensg[i, "ENSG"], split = ","))
	genes_ensg <- data.frame(PHENO = rep(cnvrs_ensg[i, "PHENO"], length(genes_ensg)),
							 CHR = rep(cnvrs_ensg[i, "CHR"], length(genes_ensg)),
							 START_CNVR = rep(cnvrs_ensg[i, "START_CNVR"], length(genes_ensg)),
							 END_CNVR = rep(cnvrs_ensg[i, "END_CNVR"], length(genes_ensg)),
							 ENSG = genes_ensg, stringsAsFactors = F)

	# Identify associated OMIM disorders
	genes_hgnc <- inner_join(genes_hgnc, morbid[, -c(3)], by = "HGNC")
	genes_ensg <- inner_join(genes_ensg, morbid[, -c(3)], by = "ENSG")

	# Overlay HGNC - ENSG results 
	genes <- unique(rbind(genes_hgnc, genes_ensg))

	# Merge to final data.frame
	cnv_morbid <- rbind(cnv_morbid, genes)
}


#################################################
### Save ########################################

fwrite(cnv_morbid, "/cauwerx/gwas/08_ANNOVAR/combined/data/OMIM_morbid_genes/CNVR_OMIM_morbid_genes_HGNC.txt", col.names = T, row.names = F, quote = F, sep = "\t")
